core/storage/statistics/gae_models_test.py

```diff
@@ -43,6 +46,125 @@ def test_create_and_get_event_models(self):
+
+    def test_insert_submitted_answers(self):
+
+        submitted_answer_list = [{'answer': 'value'}]
+
+        stat_models.StateAnswersModel.insert_submitted_answers(
+            'exp_id', 1, 'state_name', 'interaction_id',
+            submitted_answer_list)
+
+        model1 = stat_models.StateAnswersModel.get_master_model(
+            'exp_id', 1, 'state_name')
+
+        self.assertEqual(model1.exploration_id, 'exp_id')
+        self.assertEqual(model1.exploration_version, 1)
+        self.assertEqual(model1.state_name, 'state_name')
+        self.assertEqual(model1.submitted_answer_list, submitted_answer_list)
+        self.assertEqual(model1.shard_count, 0)
+
+        # Use a smaller max answer list size so fewer answers are needed to
+        # exceed a shard. This will increase the 'shard_count'.
+        with self.swap(
+            stat_models.StateAnswersModel, '_MAX_ANSWER_LIST_BYTE_SIZE', 1):
+            stat_models.StateAnswersModel.insert_submitted_answers(
+                'exp_id', 1, 'state_name', 'interaction_id',
+                submitted_answer_list)
+            stat_models.StateAnswersModel.insert_submitted_answers(
+                'exp_id', 1, 'state_name', 'interaction_id',
+                submitted_answer_list)
+
+        # 'shard_count' will not increase as number of answers are less than
+        # the max answer list size.
+        stat_models.StateAnswersModel.insert_submitted_answers(
+            'exp_id', 1, 'state_name', 'interaction_id',
+            submitted_answer_list)
+
+        model1 = stat_models.StateAnswersModel.get_master_model(
+            'exp_id', 1, 'state_name')
+
+        self.assertEqual(model1.shard_count, 2)
+
+
+class StateAnswersCalcOutputModelTests(test_utils.GenericTestBase):
+
+    def test_create_and_get_models(self):
+
+        stat_models.StateAnswersCalcOutputModel.create_or_update(
+            'exp_id', '1', 'state_name', '', 'calculation_id', '', {})
+
+        model1 = stat_models.StateAnswersCalcOutputModel.get_model(
+            'exp_id', '1', 'state_name', 'calculation_id')
+
+        self.assertEqual(model1.exploration_id, 'exp_id')
+        self.assertEqual(model1.exploration_version, '1')
+        self.assertEqual(model1.state_name, 'state_name')
+        self.assertEqual(model1.calculation_id, 'calculation_id')
+
+    def test_raise_exception_with_large_calculation_output(self):
+
+        observed_log_messages = []
+
+        def _mock_logging_function(msg, *_):
+            """Mocks logging.exception."""
+            observed_log_messages.append(msg)
+
+        logging_swap = self.swap(logging, 'exception', _mock_logging_function)
+
+        # This swaps ndb.model.put() to raise an Exception. It is required to
+        # test against large values of calculation_output.
+        put_swap = self.swap(
+            stat_models.StateAnswersCalcOutputModel, 'put',
+            lambda: (_ for _ in ()).throw(Exception()))
```

Comment:
I am not sure what's going on in this line. Could you please explain? I don't see what large values of calculation output has to do with what is written here.

This is kind of important since the test name is talking about "raise exception with large calculation output". I would expect the test to actually involve large calculation output in a proper way (i.e. without mocking).

Reply:
I am trying to test [this line](https://github.com/oppia/oppia/blob/develop/core/storage/statistics/gae_models.py#L1611) but I can't have large dicts, my terminal gets hanged or Python raises memory error. This is why I swapped the put method to raise an exception

Reply:
The size limit for datastore entities is 1 MB. Python should be able to deal with that, you can generate a string that's 1.2 MB large (say). How big are you going?

Reply:
I was doing `{i: i for i in xrange(10**x)}` and kept on increasing x till there was a memory error. But in course of increasing x, there was not a single time when an exception from put() was raised. How can I generate a dict of size 1.2 MB?

Reply:
Something like this?

```
{
    'key': 'a' * [1200000]
}
```

Interesting re the exception. I wonder whether the 1 MB limit does not apply to the dev datastore, but that seems a bit strange to me. How big did you manage to go with x?

Reply:
It works with your example, I dunno why it was not working with mine. I don't actually remember what was the value of x but I am sure it went pretty far to raise an error.